BUMP deepNOG to 1.1.0 (targeting PyPI) and update CI configurations

LokiLuciferase opened the pull request:
Just what it says on the tin. Plus, some additional refactoring and added comments.

LokiLuciferase committed:
BUMP deepNOG to 1.1.0 targeting PyPI and update CI configurations

LokiLuciferase committed:
roll back appveyor to build without using deepnog/torch since apparently torch isn't available for windows.

LokiLuciferase commented:
It is my understanding that we currently determine coverage from the AppVeyor CI build - is it easily possible to use the linux build of Travis instead @VarIr?

VarIr commented on phenotrex/structure/records.py:
```diff
@@ -1,15 +1,16 @@
+    """
+    Structure containing the genomic features to which learning is applied, for each sample
+    referenced by `identifier`.
+    """
```
```suggestion
    """
    Genomic features of a sample referenced by `identifier`.
    """
```
I'd omit learning here, as the actual feature space depends on all GenotypeRecords in the training set, right?

VarIr commented on phenotrex/structure/records.py:
```diff
@@ -19,7 +20,10 @@ def __repr__(self):
+    """
+    Structure containing ground truth class values (0 for trait absent, 1 for trait present) for the
+    trait `trait_name` in sample `identifier`.
+    """
```
```suggestion
    """
    Ground truth labels of sample `identifier`,
    indicating presence/absence of trait `trait_name`:
      - 0 if trait is absent
      - 1 if trait is present 
    """
```
Mostly for consistency to the above suggestion.

VarIr commented on phenotrex/structure/records.py:
```diff
@@ -30,20 +34,23 @@ def __repr__(self):
-    """ TODO add docstring """
+    """
+    Structure containing grouping information for each sample for Leave-one-group-out CV.
+    """
```
```suggestion
    """
    Group label of sample `identifier`.

    Notes
    -----
    Useful for leave-one-group-out cross-validation (LOGO-CV),
    for example, to take taxonomy into account.
    """
```

VarIr commented on phenotrex/structure/records.py:
```diff
@@ -30,20 +34,23 @@ def __repr__(self):
     identifier: str
-    group_name: str
-    group_id: int
+    group_name: Optional[str]
+    group_id: Optional[int]
 
     def __repr__(self):
         return f"{self.identifier} group({self.group_name})={self.group_id}"
```
When two attributes are optional, the repr will probably not look very nice.

LokiLuciferase replied:
the GroupRecord is in any case not supposed to be used outside of a training record so I modified the repr of TrainingRecord to ensure it looks okay if the group record is empty.

VarIr commented on phenotrex/structure/records.py:
```diff
@@ -30,20 +34,23 @@ def __repr__(self):
+    """
+    Structure which collates information from Genotype-, Phenotype- and GroupRecords, creating
+    a single observation suitable as machine learning input for each sample.
+    """
```
```suggestion
    """
    Sample containing Genotype-, Phenotype- and GroupRecords,
    suitable as machine learning input for a single observation.
    """
```

VarIr commented on phenotrex/transforms/annotation.py:
```diff
@@ -7,29 +8,50 @@
+class PreloadedProteinIterator(ProteinIterator):
+    """Hack ProteinDataset to load from list directly."""
+    def __init__(self, protein_list: List[SeqRecord], aa_vocab, format):
+        self.iterator = (x for x in protein_list)
```
Is that equivalent?
```suggestion
        self.iterator = protein_list
```

LokiLuciferase replied:
Yeah, should be, true

LokiLuciferase replied:
Nope, deepNOG complained about not being passed an iterator (it tried to call next() on the list)

VarIr replied:
Ah yes, the comprehension gives a generator, not a tuple, and so does not allocate memory twice. But since it's called "self.iterator", how about the following:
```suggestion
        self.iterator = iter(protein_list)
```

LokiLuciferase replied:
well technically a generator is an iterator too, but I concede the point that this is cleaner ^^

VarIr replied:
You're right. Both is fine, no hard feelings

VarIr commented on phenotrex/transforms/annotation.py:
```diff
@@ -40,7 +62,7 @@ def fastas_to_grs(fasta_files: List[str], verb: bool = False,
     :param fasta_files: a list of DNA and/or protein FASTA files to be converted into GenotypeRecords.
     :param verb: Whether to display progress of annotation with tqdm.
-    :param n_threads: Number of threads to run in parallel. Default, use up to all available CPU cores.
+    :param n_threads: Number of threads to run in parallel. Default, use all available CPU cores.
```
```suggestion
    :param n_threads: Number of parallel threads. Default, use all available CPU cores.
```

VarIr commented on tests/test_ml.py:
```diff
@@ -99,8 +96,8 @@ def test_train(self, trait_name, classifier, use_shaps):
         """
-        Test default crossvalidation of PICASVM class. Using several different traits, cv folds, and scoring methods.
-        Compares with dictionary cv_scores.
+        Test default crossvalidation of PICASVM class. Using several different traits, cv folds,
```
PICASVM? Never heard of that ^^

VarIr commented on phenotrex/util/plotting.py:
```diff
@@ -5,7 +5,12 @@
 
 
-def compleconta_plot(cccv_results: List[Dict[float, Dict[float, Dict[str, float]]]],
+CccvType = Union[
+    Dict[float, Dict[float, Dict[str, float]]],
+    List[Dict[float, Dict[float, Dict[str, float]]]]
+]
```
Wild :)

LokiLuciferase replied:
yeah, right? :grin:

VarIr reviewed (approved):
Thanks for the PR! I see no blockers here, just some minor comments.

VarIr reviewed (changes requested):
Just saw that all Travis builds failed. While the Linux builds fail due to the same issue we see in deepnog, the mac builds fail due to something wrong with prodigal. These should be investigated.

VarIr commented:
About coverage: It seems to me, we actually run codecov on Travis, but not on AppVeyor? In any case, it should be simple to enable it on any platform. codecov claim, they automatically merge reports from different platforms. So perhaps just add a line "codecov" to `.appveyor.yml` after the tests (and `pytest --cov`, unless we already have it).

VarIr commented:
Btw, see the [deepnog .appveyor.yml](https://github.com/univieCUBE/deepnog/blob/master/.appveyor.yml#L54) for how to obtain `torch` on Windows. (Hasn't found its way into setup.py, yet).

LokiLuciferase commented:
> Just saw that all Travis builds failed. While the Linux builds fail due to the same issue we see in deepnog, the mac builds fail due to something wrong with prodigal. These should be investigated.

Yeah, it was possible that the prodigal binary compiled for linux wouldn't run under macos (and it sure as hell won't under Win10). I can look into having it compile from source under linux/macos, and ship the windows binary with the repo.

LokiLuciferase committed:
address comments by @VarIr

LokiLuciferase committed:
ADD prodigal binaries for win32 and darwin, and select appropriate at inference time

LokiLuciferase committed:
switch license to GPLv3 due to inclusion of prodigal binaries

LokiLuciferase committed:
FIX PRODIGAL_BIN_PATH and revert changes to PreloadedProteinIterator

LokiLuciferase committed:
UPDATE call_proteins function to use pre-loaded DNA fasta sequences (distributed prodigal binaries don't handle gzipped files)

LokiLuciferase committed:
increase test coverage for io module

LokiLuciferase committed:
ADD taxid groups file to increase test coverage for ml module

LokiLuciferase commented:
- I have now added all prodigal binaries for linux, win32 and darwin architectures. I had to adapt some of the protein calling code since the pre-compiled packages don't allow passing gzipped files. 
- I also changed the license of phenotrex to GPL v3 to comply with the license of prodigal.
- Finally, I added some more tests for file I/O and added a groups file for Sulfate_reducer to be able to test LOGO CV.

VarIr commented on LICENSE:
```diff
@@ -1,22 +1,15 @@
-MIT License
+phenotrex
+Copyright (C) 2019, Lukas Lüftinger
```
```suggestion
Copyright (C) 2019-2020, Lukas Lüftinger
```

VarIr commented on phenotrex/structure/records.py:
```diff
@@ -1,15 +1,15 @@
 class GenotypeRecord:
-    """ TODO add docstring """
+    """
+    enomic features of a sample referenced by `identifier`.
```
I buy a G
```suggestion
    Genomic features of a sample referenced by `identifier`.
```

VarIr commented on phenotrex/transforms/annotation.py:
```diff
@@ -66,43 +90,53 @@ def fasta_to_gr(fasta_file: str, verb: bool = False) -> GenotypeRecord:
+    parsed = list(parse(faa_file, 'fasta'))
+    faa_file.close()
+    os.unlink(fna_file.name)
+    os.unlink(faa_file.name)  # cannot re-use .name of open NamedTemporaryFile under Win32
```
Why not use a `with` context manager?

VarIr reviewed (approved):
It's good to see the coverage go up! I'd prefer to see the gene calling in a context manager, if possible, but otherwise LGTM.

VarIr commented:
Btw, the Travis errors will likely go away with PyTorch 1.5. We might want to allow Travis to fail the Linux builds until then (or simply ignore the errors).

LokiLuciferase committed:
address comments by @VarIr

LokiLuciferase commented:
Merging this now. @VarIr maybe you can upload to pypi when you find time?

VarIr commented:
Thank you for the PR! I've just created a new release and uploaded to PyPI.
